src/shims/tsd.h

```diff
@@ -92,10 +92,75 @@ _dispatch_thread_key_create(const unsigned long *k, void (*d)(void *))
 	if (!*k || !d) return;
 	dispatch_assert_zero(pthread_key_init_np((int)*k, d));
 }
+#elif DISPATCH_USE_THREAD_LOCAL_STORAGE
+
+DISPATCH_TSD_INLINE
+static inline void
+_dispatch_thread_key_create(pthread_key_t *k, void (*d)(void *))
+{
+	dispatch_assert_zero(pthread_key_create(k, d));
+}
+
+struct dispatch_tsd {
+	bool initialized;
+	pid_t tid;
+	void *dispatch_queue_key;
+	void *dispatch_frame_key;
+	void *dispatch_cache_key;
+	void *dispatch_context_key;
+	void *dispatch_pthread_root_queue_observer_hooks_key;
+	void *dispatch_defaultpriority_key;
+#if DISPATCH_INTROSPECTION
+	void *dispatch_introspection_key;
+#elif DISPATCH_PERF_MON
+	void *dispatch_bcounter_key;
+#endif
+#if DISPATCH_LOCK_USE_SEMAPHORE_FALLBACK
+	void *dispatch_sema4_key;
+#endif
+	void *dispatch_priority_key;
+	void *dispatch_voucher_key;
+	void *dispatch_deferred_items_key;
+};
+
+extern __thread struct dispatch_tsd __dispatch_tsd;
+extern pthread_key_t __dispatch_tsd_key;
+extern void libdispatch_tsd_init(void);
+
+DISPATCH_ALWAYS_INLINE
```

Comment:
you should also add `DISPATCH_PURE` here, which can help have the compiler hoist the lazy init to the top.